man/mcp-sparse.1
.\" Copyright (c) 2019-2021 Miki Hermann & Gernot Salzer
.TH mcp-sparse 1 "2021-03-19" "1.04" "MCP System"
.
.SH NAME
mcp-sparse - transforms a file of vectors in sparse format into the
full format
.
.SH SYNOPSIS
.B mcp-sparse
.RI [\| "OPTION" "\|]\|.\|.\|."
.
.SH DESCRIPTION
.PP
Transforms a file with vectors in sparse format prefixed by
group identifiers, into fully unfolded vectors.
.PP
The sparse format is a notation for sparse vectors of the form
.IP
.in +4n
.EX
.IR "STRING " [\| INTEGER " : " STRING \|]\|.\|.\|.
.EE
.in
.PP
where the first \fISTRING\fR is the group prefix and the pairs
\fIINTEGER\fR : \fISTRING\fR denote the coordinates and its values.
The pairs \fIINTEGER\fR : \fISTRING\fR must be listed in increasing
order.
.
.SH OPTIONS
.
.TP
\fB\-i\fR, \fB\-\-input\fI input-file
Input file containing vectors in sparse format prefixed by group identifiers.
.IP
Default: STDIN.
.
.TP
\fB\-o\fR, \fB\-\-output\fI output-file
Output file, containing vectors in full format. If
there is no suffix starting with a dot, the software adds \fI.data\fR
to the \fIinput-file\fR.
.IP
Default: STDOUT or the \fIbasename\fR of \fIinput-file\fR with \fI.data\fR suffix.
.
.TP
\fB\-d\fR, \fB\-\-default\fI STRING
Default \fISTRING\fR which will be filled in for missing values.
.IP
Default: 0.
.
.
.SH SEE ALSO
mcp-guess(1),
mcp-trans(1),
mcp-split(1),
mcp-seq(1),
mcp-mpi(1),
mcp-pthread(1),
mcp-hybrid(1),
mcp-check(1),
mcp-chk2tst(1),
mcp-predict(1)
.
.SH BUGS
There are certanly some.
.
.SH AUTHORS
Miki Hermann <hermann@lix.polytechnique.fr>
.br
Gernot Salzer <gernot.salzer@tuwien.ac.at>


changelog.txt
1.04
- fixed a bug in mcp-check.cpp
- added LaTeX output
- mcp-split introduced
- post_prod introduced
- write formula to a file (DIMACS form)

1.03c
 - added print option DIMACS
 - corrected sorting bug when '--cooking raw'
 - corrected bug when '--strategy exact' and '--closure Horn'
 - index of group_of_matrix: int -> string
 - variables can be named in the matrix
 - transformation in C++
 - failure to open input or output file results in exit and not in exception
 - added options 'peek' and 'section' for printing matrices
 - when variables have own names, literals made readable
 - bug correction: 'continue' instead of 'return' when empty formula
   in sequential for bijunctive 
 - added mixed printing of formulas
 - added resolution with binary clause which contracts the target clause
   (l0 + l1) * (l0 + -l1 + c) => (l0 + l1) * (l0 + c)
 - added mcp-cnf producting exact CNF formula


src.bak/mcp-sparse.cpp
/**************************************************************************
 *                                                                        *
 *                                                                        *
 *	         Multiple Classification Project (MCP)                    *
 *                                                                        *
 *	Author:   Miki Hermann                                            *
 *	e-mail:   hermann@lix.polytechnique.fr                            *
 *	Address:  LIX (CNRS UMR 7161), Ecole Polytechnique, France        *
 *                                                                        *
 *	Author: Gernot Salzer                                             *
 *	e-mail: gernot.salzer@tuwien.ac.at                                *
 *	Address: Technische Universitaet Wien, Vienna, Austria            *
 *                                                                        *
 *	Version: all                                                      *
 *      File:    mcp-sparse.cpp                                           *
 *                                                                        *
 *      Copyright (c) 2019 - 2025                                         *
 *                                                                        *
 * Transform files from sparse format to full format                      *
 *                                                                        *
 **************************************************************************/

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include "mcp-matrix+formula.hpp"

using namespace std;

#define STDIN    "STDIN"
#define STDOUT   "STDOUT"

string input   = STDIN;
string output  = STDOUT;
string dflt    = "0";

ifstream infile;
ifstream metafile;
ofstream outfile;
streambuf *backup;

int max_idx = 0;

//------------------------------------------------------------------------------

bool isblank(const string &id) {
  if (id.empty())
    return true;
  for (int i = 0; i < id.length(); ++i)
    if (isspace(id[i]) || iscntrl(id[i]) || !isprint(id[i]))
      return true;
  return false;
}

void read_arg  (int argc, char *argv[]) {	// reads the input parameters
  int argument = 1;
  while (argument < argc) {
    string arg = argv[argument];
    if (arg == "-i"
	|| arg == "--input") {
      input = argv[++argument];
    } else if (arg == "-o"
	       || arg == "--output") {
      output = argv[++argument];
    } else if (arg == "-d"
	       || arg == "--default") {
      dflt = argv[++argument];
    } else {
    cerr << "+++ argument error: " << arg << " " << argv[argument] << endl;
    exit(1);
    }
    ++argument;
  }
  if (input != STDIN && output == STDOUT) {
    string::size_type pos = input.rfind('.');
    output = (pos == string::npos ? input : input.substr(0, pos)) + ".data";
  }
  if (isblank(dflt)) {
    cerr << "*** Default string cannot be white space or control char" << endl;
    exit(1);
  }
}

void IO_open () {
  if (input != STDIN) {
    infile.open(input);
    if (infile.is_open()) {
      cin.rdbuf(infile.rdbuf());
    } else {
      cerr << "+++ Cannot open input file " << input << endl;
      exit(1);
    }
  }
  
  if (output != STDOUT) {
    outfile.open(output);
    if (outfile.is_open()) {
      backup = cout.rdbuf();
      cout.rdbuf(outfile.rdbuf());
    } else {
      cerr << "+++ Cannot open output file " << output << endl;
      exit(1);
    }
  }
}

void IO_close () {
  if (input != STDIN)
    infile.close();
  if (output != STDOUT) {
    outfile.close();
    cout.rdbuf(backup);
  }
}

void I_reopen () {
  if (input != STDIN) {
    infile.close();
    infile.open(input);
    if (infile.is_open()) {
      cin.rdbuf(infile.rdbuf());
    } else {
      cerr << "+++ Cannot reopen input file " << input << endl;
      exit(1);
    }
  }
}

void spread (const string &line) {
  vector<string> parts = split(line, " \t");
  cout << parts[0];

  int ptr = 1;
  for (int i = 1; i < parts.size(); ++i) {
    vector<string> pair = split(parts[i], ":");
    
    while (ptr < stoi(pair[0])) {
      cout << " " << dflt;
      ptr++;
    }
    cout << " " << pair[1];
    ptr++;
  }
  while (ptr <= max_idx) {
    cout << " " << dflt;
    ptr++;
  }
  cout << endl;
}

//////////////////////////////////////////////////////////////////////////////

int main(int argc, char **argv)
{
  string line;
  vector<string> sparse_data;

  version += "sparse";
  cerr << "+++ version = " << version << endl;

  read_arg(argc, argv);
  IO_open();

  while (getline(cin, line)) {
    if (input == STDIN)
      sparse_data.push_back(line);

    vector<string> parts = split(line, " \t");
    vector<string> last = split(parts[parts.size()-1], ":");
    max_idx = max(max_idx, stoi(last[0]));
  }

  I_reopen();

  if (input == STDIN)
    for (int ln = 0; ln < sparse_data.size(); ++ln)
      spread(sparse_data[ln]);
  else
    while (getline(cin, line))
      spread(line);
  
  IO_close();
  return 0;
}

////////////////////////////////////////////////////////////////////////////////
